src/main/java/org/truffleruby/cext/CExtNodes.java

```diff
@@ -1035,6 +1038,47 @@ protected Object rbConstSet(RubyModule module, String name, Object value,
 
     }
 
+    @Primitive(name = "rb_gv_get")
+    @NodeChild(value = "name", type = RubyNode.class)
+    public abstract static class RbGvGetNode extends PrimitiveNode {
+
+        @Child WarnNode warnNode;
+
+        @CreateCast("name")
+        protected RubyNode coerceToString(RubyNode name) {
+            return ToJavaStringNode.create(name);
+        }
+
+        @Specialization
+        protected Object rbGvGet(String name,
+                @Cached BranchProfile notExistsProfile) {
+            boolean exists = getContext().getCoreLibrary().globalVariables.contains(name);
+
+            if (!exists) {
+                notExistsProfile.enter();
+                warn(StringUtils.format("global variable `%s' not initialized", name));
+
+                return nil;
+            }
+
+            return getContext().getCoreLibrary().globalVariables.getReader(name).getValue();
```

Comment:
It's completely fine to add a spec checking both and how they interact with each other, simply put it together with any of these 2 function specs. Just like we always need to use the setter method when testing some getter-like method.
Just testing `rb_define_readonly_variable()` with some non-nil value and then `rb_gv_get()` should show the issue.

There is already a spec for rb_define_readonly_variable, so you could just test reading `$ro_gvar` using rb_gv_get in addition to reading it from Ruby code directly.

Reply:
Even making org.truffleruby.language.globals.ReadGlobalVariableNode usable with a variable name wouldn't help, because that uses LookupGlobalVariableStorageNode and that creates the storage eagerly. The whole design of global variable storage is based around creating it eagerly so we can cache it. And due to global variable aliases we can't safely replace an existing storage for a name, and we assume it doesn't changed unless aliased.

I guess the only thing left is to have a static method in ReadGlobalVariableNode that replicates the logic but is entirely uncached and looks the storage without creating it eagerly.

Reply:
In this case, I bail early if the variable doesn't exist. Unless you dislike the call to `getContext().getCoreLibrary().globalVariables.contains(javaName)`, I think we can use `ReadGlobalVariableNode`, with some new specialization that takes a String name. By that point in `rbGvGet`, the global variable must already be defined in storage.

Reply:
Do you mean adding a `String name` parameter for `org.truffleruby.language.globals.ReadGlobalVariableNode`?
That seems difficult, literally all the code there assumes a fixed `name` (it's a final field of the node).

If you mean `org.truffleruby.core.kernel.TruffleKernelNodes.ReadGlobalVariableNode` then it's the same problem, as that uses `org.truffleruby.language.globals.ReadGlobalVariableNode`. So it'd only work if `org.truffleruby.language.globals.ReadGlobalVariableNode` can be made `@GenerateUncached` but that seems again quite difficult with the `name` field.

I see the approach you pushed. That works by using `eval` for the uncached case, but programs might run in `eval` very quickly. Notably, there is nothing I can see which triggers splitting here, so we'll get to `rbGvGetUncached()` as soon as `rb_gv_get()` has been called globally with 9 different variable names. So I guess we could `@ReportPolymorhism` on `rb_gv_get()` as a workaround (but that won't work with Mode=latency or Splitting disabled).
I think ultimately we need to solve this so it's less expensive than an `eval` for the uncached case, but that doesn't need to be solved in this PR.

Reply:
I forgot, it was already `eval` so it's not worse.
Still could you add `@ReportPolymorhism` so at least we have less chances to get to the uncached case?

Reply:
I've added the `@ReportPolymorphism` annotation. Agreed that without splitting this devolves into a slow path. As you've noted, we were using `eval` 100% of the time previously, so this shouldn't be any worse. I'm not a fan of using `eval` here, but changing `ReadGlobalVariableNode` or some other take a String at runtime was also quite ugly. We'll want to take another pass over this for sure.